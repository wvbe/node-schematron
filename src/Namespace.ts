export class Namespace {
	prefix: string;
	uri: string;

	constructor(prefix: string, uri: string) {
		this.prefix = prefix;
		this.uri = uri;
	}

	static QUERY = `map {
		"prefix": @prefix/string(),
		"uri": @uri/string()
	}`;

	static fromJson(json: NamespaceJson): Namespace {
		return new Namespace(json.prefix, json.uri);
	}
}

export type NamespaceJson = {
	prefix: string;
	uri: string;
};
