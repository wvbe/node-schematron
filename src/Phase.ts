import { Variable, VariableJson } from './Variable';

export class Phase {
	id: string;
	active: string[];
	variables: Variable[];

	constructor(id: string, active: string[], variables: Variable[]) {
		this.id = id;
		this.active = active;
		this.variables = variables;
	}

	static QUERY = `map {
		"id": @id/string(),
		"active": array { ./sch:active/@pattern/string() },
		'variables': array { ./sch:let/${Variable.QUERY}}
	}`;

	static fromJson(json: PhaseJson): Phase {
		return new Phase(
			json.id,
			json.active,
			json.variables.map(rule => Variable.fromJson(rule))
		);
	}
}

export type PhaseJson = {
	id: string;
	active: string[];
	variables: VariableJson[];
};
