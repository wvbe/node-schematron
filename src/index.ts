export { registerCustomXPathFunction } from 'fontoxpath';

export { Schema } from './Schema';
