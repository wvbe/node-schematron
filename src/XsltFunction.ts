import { Namespace, NS_XSLT } from './Namespace';
import { Schema } from './Schema';
import {
	instantiateSequenceConstructorFromJson,
	joinSequenceConstructors,
	XsltSequenceConstructor,
	XsltSequenceConstructorChoose,
	XsltSequenceConstructorForEach,
	XsltSequenceConstructorIf,
	XsltSequenceConstructorJson,
	XsltSequenceConstructorValueOf,
	XsltSequenceConstructorVariable
} from './XsltSequenceConstructor';

type XsltDataType = 'xs:boolean' | 'xs:string';

type XsltFunctionParam = {
	type: XsltDataType;
	name: string | null;
};

export type XsltFunctionJson = {
	name: string;
	parameters: XsltFunctionParam[];
	sequenceConstructors: XsltSequenceConstructorJson[];
	returnValue: XsltDataType;
};

export class XsltFunction {
	localName: string;
	namespaceUri?: string;
	namespacePrefix?: string;
	parameters: XsltFunctionParam[];
	sequenceConstructors: XsltSequenceConstructor[];
	returnValue: XsltDataType;

	constructor(
		name: string,
		parameters: XsltFunctionParam[],
		// https://www.w3.org/TR/xslt20/#dt-sequence-constructor
		sequenceConstructors: XsltSequenceConstructor[],
		returnValue: XsltDataType
	) {
		const fullyQualifiedMatch = /Q{(.*)}(.*)/.exec(name);
		if (fullyQualifiedMatch) {
			this.namespaceUri = fullyQualifiedMatch[1];
		} else if (name.includes(':')) {
			this.namespacePrefix = name.substr(0, name.indexOf(':'));
		} else {
			throw new Error('Custom functions must have a namespace');
		}

		this.localName = name;
		this.parameters = parameters;
		this.sequenceConstructors = sequenceConstructors;
		this.returnValue = returnValue;
	}

	// This method converts an <xsl:function> into an XQuery function definition. That definition is later thrown into
	// an XQuery (library) module that correlates with the namespace.
	getXqueryDefinition(schema: Schema): string {
		const parameters = this.parameters
			.map(param => `\n\t$${param.name} as ${param.type || 'item()'}`)
			.join(',');
		const schematronVariables = schema.variables.map(
			variable =>
				new XsltSequenceConstructorVariable({
					type: 'variable',
					children: [],
					name: variable.name,
					select: variable.value
				})
		);
		const sequenceConstructors = joinSequenceConstructors([
			...schematronVariables,
			...this.sequenceConstructors
		] as XsltSequenceConstructor[]);
		return `declare %public function ${this.localName}(${parameters}\n) {${sequenceConstructors}\n};`;
	}

	isInNamespace(namespace: Namespace): boolean {
		if (this.namespaceUri) {
			return namespace.uri === this.namespaceUri;
		}
		if (this.namespacePrefix) {
			return namespace.prefix === this.namespacePrefix;
		}
		return false;
	}

	static FUNCTIONS = `declare function local:sequence-constructors($node as node()*) {
		array {
			(: TODO all possible sequence constructors :)
			$node/Q{${NS_XSLT}}variable/${XsltSequenceConstructorVariable.QUERY},
			$node/Q{${NS_XSLT}}value-of/${XsltSequenceConstructorValueOf.QUERY},
			$node/Q{${NS_XSLT}}choose/${XsltSequenceConstructorChoose.QUERY},
			$node/Q{${NS_XSLT}}if/${XsltSequenceConstructorIf.QUERY},
			$node/Q{${NS_XSLT}}for-each/${XsltSequenceConstructorForEach.QUERY}
		}
	};`;

	static QUERY = `map {
		"name": @name/string(),
		"returnValue": @as/string(),
		"parameters": array { ./Q{${NS_XSLT}}param/map {
			"name": @name/string(),
			(:
				If the as attribute is omitted, then the required type is item()*.
				https://www.w3.org/TR/xslt-30/#parameter-type
			:)
			"type": @as/string()
		}},
		"sequenceConstructors": local:sequence-constructors(.)
	}`;

	static fromJson(json: XsltFunctionJson): XsltFunction {
		return new XsltFunction(
			json.name,
			json.parameters,
			json.sequenceConstructors
				.map(instantiateSequenceConstructorFromJson)
				.filter(Boolean) as XsltSequenceConstructor[],
			json.returnValue
		);
	}
}
