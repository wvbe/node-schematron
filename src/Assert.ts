import { evaluateXPathToBoolean, evaluateXPathToString } from 'fontoxpath';

import { Result } from './Result';

import { FontoxpathOptions } from './types';

export class Assert {
	id: string | null;
	test: string;
	message: Array<string | Object>;
	isReport: boolean;

	constructor(
		id: string | null,
		test: string,
		message: Array<string | Object>,
		isReport: boolean
	) {
		this.id = id;
		this.test = test;
		this.message = message;
		this.isReport = isReport;
	}

	createMessageString(
		contextNode: Node,
		variables: Object,
		fontoxpathOptions: FontoxpathOptions,
		chunks: Array<string | any>
	): string {
		return chunks
			.map((chunk): string => {
				if (typeof chunk === 'string') {
					return chunk;
				}

				// <sch:name />
				if (chunk.$type === 'name') {
					return evaluateXPathToString(
						'name(' + (chunk.path || '') + ')',
						contextNode,
						null,
						variables,
						fontoxpathOptions
					);
				}

				// <sch:value-of />
				if (chunk.$type === 'value-of') {
					return evaluateXPathToString(
						chunk.select,
						contextNode,
						null,
						variables,
						fontoxpathOptions
					);
				}

				console.log(chunk);
				throw new Error('Unsupported element in <sch:message>');
			})
			.join('');
	}

	validateNode(
		context: Node,
		variables: Object,
		fontoxpathOptions: FontoxpathOptions
	): Result | null {
		const outcome = evaluateXPathToBoolean(
			this.test,
			context,
			null,
			variables,
			fontoxpathOptions
		);
		return (!this.isReport && outcome) || (this.isReport && !outcome)
			? null
			: new Result(
					context,
					this,
					this.createMessageString(context, variables, fontoxpathOptions, this.message)
			  );
	}

	static QUERY = `map {
		'id': if (@id) then string(@id) else (),
		'test': @test/string(),
		'message': array { (./text()|./element())/local:json(.) },
		'isReport': boolean(local-name() = 'report')
	}`;

	static fromJson(json: AssertJson): Assert {
		return new Assert(json.id, json.test, json.message, json.isReport);
	}
}

export type AssertJson = {
	id: string | null;
	test: string;
	message: Array<string | Object>;
	isReport: boolean;
};
