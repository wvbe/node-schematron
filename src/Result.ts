import { Node, serializeToWellFormedString } from 'slimdom';
import { Assert } from './Assert';

export class Result {
	// pattern: Pattern;
	// phase?: Phase;
	// rule: Rule;
	assertId: string | null;
	isReport: boolean;
	context: Node;
	message?: string;

	constructor(
		// pattern: Pattern,
		// phase?: Phase,
		// rule: Rule,
		context: Node,
		assert: Assert,
		message?: string
	) {
		// this.pattern = pattern;
		// this.phase = phase;
		// this.rule = rule;
		this.assertId = assert.id;
		this.isReport = assert.isReport;
		this.context = context;
		this.message = message;
	}

	toJson() {
		return {
			assertId: this.assertId,
			isReport: this.isReport,
			context: serializeToWellFormedString(this.context),
			message: this.message
		};
	}
}
