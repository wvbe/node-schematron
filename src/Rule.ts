import { Assert, AssertJson } from './Assert';
import { Result } from './Result';
import { Variable, VariableJson } from './Variable';

import { FontoxpathOptions } from './types';

export class Rule {
	context: string;
	variables: Variable[];
	asserts: Assert[];

	constructor(context: string, variables: Variable[], asserts: Assert[]) {
		this.context = context;
		this.variables = variables;
		this.asserts = asserts;
	}

	validateNode(
		context: Node,
		parentVariables: Object | null,
		fontoxpathOptions: FontoxpathOptions
	): Result[] {
		const variables = Variable.reduceVariables(context, this.variables, fontoxpathOptions, {
			...parentVariables
		});

		return this.asserts
			.map(assert => assert.validateNode(context, variables, fontoxpathOptions))
			.filter(result => result !== null) as Result[];
	}

	static QUERY = `map {
		'context': @context/string(),
		'variables': array { ./sch:let/${Variable.QUERY}},
		'asserts': array{ ./(sch:report|sch:assert)/${Assert.QUERY}}
	}`;

	static fromJson(json: RuleJson): Rule {
		const variables = json.variables.map(rule => Variable.fromJson(rule));
		const asserts = json.asserts.map(rule => Assert.fromJson(rule));

		return new Rule(json.context, variables, asserts);
	}
}

export type RuleJson = {
	context: string;
	variables: VariableJson[];
	asserts: AssertJson[];
};
