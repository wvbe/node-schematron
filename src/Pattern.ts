import { evaluateXPathToNodes } from 'fontoxpath';

import { Result } from './Result';
import { Rule, RuleJson } from './Rule';
import { Variable, VariableJson } from './Variable';

import { FontoxpathOptions } from './types';

export class Pattern {
	id: string | null;
	rules: Rule[];
	variables: Variable[];

	constructor(id: string | null, rules: Rule[], variables: Variable[]) {
		this.id = id;
		this.rules = rules;
		this.variables = variables;
	}

	validateDocument(
		documentDom: Document,
		parentVariables: object | null,
		fontoxpathOptions: FontoxpathOptions
	) {
		const variables = Variable.reduceVariables(documentDom, this.variables, fontoxpathOptions, {
			...parentVariables
		});
		const ruleContexts = this.rules.map(rule =>
			evaluateXPathToNodes(
				'//(' + rule.context + ')',
				documentDom,
				null,
				variables,
				fontoxpathOptions
			)
		);
		const flattenValidationResults = (results: Result[], node: Node): Result[] => {
			const ruleIndex = ruleContexts.findIndex(context => context.includes(node));
			const rule = ruleIndex >= 0 ? this.rules[ruleIndex] : null;
			if (rule) {
				results.splice(
					results.length,
					0,
					...rule.validateNode(node, variables, fontoxpathOptions)
				);
			}

			return Array.from(node.childNodes).reduce(flattenValidationResults, results);
		};

		return Array.from(documentDom.childNodes).reduce(flattenValidationResults, []);
	}

	static QUERY = `map {
		'id': @id/string(),
		'rules': array{ ./sch:rule/${Rule.QUERY}},
		'variables': array { ./sch:let/${Variable.QUERY}}
	}`;

	static fromJson(json: PatternJson): Pattern {
		return new Pattern(
			json.id,
			json.rules.map(obj => Rule.fromJson(obj)),
			json.variables.map(obj => Variable.fromJson(obj))
		);
	}
}

export type PatternJson = {
	id: string | null;
	rules: RuleJson[];
	variables: VariableJson[];
};
