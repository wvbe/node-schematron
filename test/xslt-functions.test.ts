import { evaluateXPath } from 'fontoxpath';
import { sync } from 'slimdom-sax-parser';
import { Schema } from '../src/index';
import { XsltFunction } from '../src/XsltFunction';

const xml = `<xml><num>0.5</num><num>1.5</num></xml>`;

describe('Custom XPath functions', () => {
	// it('Parses to the expected AST', () => {
	// 	expect(schema).toMatchSnapshot();
	// });

	xit('Work in reports and asserts', () => {
		const schema = Schema.fromString(
			`<schema xmlns="http://purl.oclc.org/dsdl/schematron">
				<ns uri="test" prefix="test" />
				<function xmlns="http://www.w3.org/1999/XSL/Transform" name="test:one" as="xs:boolean">
					<param name="input" as="xs:decimal"/>
					<value-of select="$input &gt; 1"/>
				</function>
				<pattern>
					<rule context="num">
						<report test="test:one(number(.))">Greater than one</report>
						<assert test="test:one(number(.))">Not greater than one</assert>
					</rule>
				</pattern>
			</schema>`
		);

		const results = schema.validateString(xml, { debug: true });
		expect(results).toHaveLength(2);
		expect(results[0].message).toBe('Not greater than one');
		expect(results[1].message).toBe('Greater than one');
	});

	describe('Convertig XSLT to XQUF', () => {
		it('Simple function', () => {
			const json = evaluateXPath(
				XsltFunction.QUERY,
				sync(`<function xmlns="http://www.w3.org/1999/XSL/Transform" name="test:one" as="xs:boolean">
					<param name="input" as="xs:decimal"/>
					<value-of select="$input &gt; 1"/>
				</function>`).documentElement
			);
			expect(json).toEqual({
				name: 'test:one',
				parameters: [
					{
						name: 'input',
						type: 'xs:decimal'
					}
				],
				returnValue: 'xs:boolean',
				sequenceConstructors: [{ type: 'value-of', select: '$input > 1' }]
			});
			expect(XsltFunction.fromJson(json).getXqueryDefinition()).toMatchSnapshot();
		});
		it('PEPPOL function using variables', () => {
			const json = evaluateXPath(
				XsltFunction.QUERY,
				sync(`<function xmlns="http://www.w3.org/1999/XSL/Transform" name="u:gln" as="xs:boolean">
					<param name="val"/>
					<variable name="length" select="string-length($val) - 1"/>
					<variable name="digits" select="reverse(for $i in string-to-codepoints(substring($val, 0, $length + 1)) return $i - 48)"/>
					<variable name="weightedSum" select="sum(for $i in (0 to $length - 1) return $digits[$i + 1] * (1 + ((($i + 1) mod 2) * 2)))"/>
					<value-of select="(10 - ($weightedSum mod 10)) mod 10 = number(substring($val, $length + 1, 1))"/>
				</function>`).documentElement
			);
			expect(json).toEqual({
				name: 'u:gln',
				parameters: [
					{
						name: 'val',
						type: null
					}
				],
				returnValue: 'xs:boolean',
				sequenceConstructors: [
					{
						type: 'variable',
						name: 'length',
						select: 'string-length($val) - 1'
					},
					{
						type: 'variable',
						name: 'digits',
						select:
							'reverse(for $i in string-to-codepoints(substring($val, 0, $length + 1)) return $i - 48)'
					},
					{
						type: 'variable',
						name: 'weightedSum',
						select:
							'sum(for $i in (0 to $length - 1) return $digits[$i + 1] * (1 + ((($i + 1) mod 2) * 2)))'
					},
					{
						type: 'value-of',
						select:
							'(10 - ($weightedSum mod 10)) mod 10 = number(substring($val, $length + 1, 1))'
					}
				]
			});
			console.log(XsltFunction.fromJson(json).getXqueryDefinition());
			expect(XsltFunction.fromJson(json).getXqueryDefinition()).toMatchSnapshot();
		});
	});
});
