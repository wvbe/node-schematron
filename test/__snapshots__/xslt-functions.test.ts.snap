// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom XPath functions Convertig XSLT to XQUF JUDE simple function using xsl:choose 1`] = `
"declare %public function local-fn:countAgeID(
	$ID as xs:string
) {
  (: CHOOSE :)
  if (($Persons[@s:id=$ID]/nc:PersonAgeMeasure))
    then (
    )
    else (
      (0)
    )
};"
`;

exports[`Custom XPath functions Convertig XSLT to XQUF PEPPOL function using variables 1`] = `
"declare %public function u:gln(
	$val as item()
) {
  let $length := string-length($val) - 1
  let $digits := reverse(for $i in string-to-codepoints(substring($val, 0, $length + 1)) return $i - 48)
  let $weightedSum := sum(for $i in (0 to $length - 1) return $digits[$i + 1] * (1 + ((($i + 1) mod 2) * 2)))
  return (
    ((10 - ($weightedSum mod 10)) mod 10 = number(substring($val, $length + 1, 1)))
  )
};"
`;

exports[`Custom XPath functions Convertig XSLT to XQUF Simple function 1`] = `
"declare %public function test:one(
	$input as xs:decimal
) {
  ($input > 1)
};"
`;

exports[`Custom XPath functions Convertig XSLT to XQUF xsl:choose 1`] = `
"declare %public function local:test(
) {
  (: CHOOSE :)
  if (X)
    then (
    )
  else if (X)
    then (
    )
    else (
      (X)
    )
};"
`;

exports[`Custom XPath functions Convertig XSLT to XQUF xsl:for-each 1`] = `
"declare %public function local-fn:checkPresent(
	$find as xs:string,
	$list as xs:string
) {
  (0),
  for $tweak_me in tokenize($find,',')
    return $tweak_me/(
      if (matches(., $list))
        then 
          (1)
        else ()
    )
};"
`;
